vortex-serde/src/layout/reader/mod.rs

```diff
@@ -0,0 +1,317 @@
+                                let f = async move {
+                                    let ar = read;
+                                    let bufs = stream::iter(r1.into_iter())
+                                        .map(|(id, range)| {
+                                            println!("Range id: {id:?}, bytes: {range:?}");
+                                            let mut buf = BytesMut::with_capacity(range.len());
+                                            unsafe { buf.set_len(range.len()) }
+                                            let buf_result = ar.read_at_into(range.begin, buf);
+                                            buf_result.map(|res_ft| {
+                                                res_ft
+                                                    .map(|res| (id, res.freeze()))
+                                                    .map_err(VortexError::from)
+                                            })
+                                        })
+                                        .buffered(10)
+                                        .try_collect()
+                                        .await;
+                                    bufs.map(|b| (ar, b))
+                                }
+                                .boxed();
+                                self.state = StreamingState::Reading(f);
+                            }
+                            ReadResult::Batch(a) => self.state = StreamingState::Decoding(a),
+                        }
+                    } else {
+                        return Poll::Ready(None);
+                    }
+                }
+                StreamingState::Decoding(arr) => {
+                    let mut batch = arr.clone();
+                    if self.scan.indices.is_some() {
+                        batch = self.take_batch(&batch)?;
+                    }
+
+                    if let Some(row_filter) = &self.scan.filter {
+                        let mask = filter_indices(&batch, &row_filter.disjunction)?;
+                        batch = filter(&batch, &mask)?;
+                    }
+
+                    let projected = match &self.scan.projection {
+                        Projection::All => batch,
+                        Projection::Partial(indices) => StructArray::try_from(batch.clone())?
+                            .project(indices.as_ref())?
+                            .into_array(),
+                    };
+
+                    self.state = StreamingState::Init;
+                    return Poll::Ready(Some(Ok(projected)));
+                }
+                StreamingState::Reading(f) => match ready!(f.poll_unpin(cx)) {
+                    Ok((read, buffers)) => {
+                        let mut write_cache = self.messages_cache.write().unwrap();
+                        for (id, buf) in buffers {
+                            write_cache.set(id, buf)
+                        }
+                        drop(write_cache);
```

Comment:
Otherwise compiler things that the cache lock is still in use and you get conflicting mutable and immutable borrows. However, since we are done with the cache we can safely drop it